src/source/PeerConnection/Retransimitter.c

```diff
@@ -0,0 +1,124 @@
+    pRetransmitter->validIndexList = MEMALLOC(SIZEOF(UINT64) * validIndexListLen);
+    pRetransmitter->validIndexListLen = validIndexListLen;
+
+CleanUp:
+    if (STATUS_FAILED(retStatus) && pRetransmitter != NULL) {
+        freeRetransmitter(&pRetransmitter);
+        pRetransmitter = NULL;
+    }
+    if (ppRetransmitter != NULL) {
+        *ppRetransmitter = pRetransmitter;
+    }
+    LEAVES();
+    return retStatus;
+}
+
+STATUS freeRetransmitter(PRetransmitter* ppRetransmitter)
+{
+    ENTERS();
+
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(ppRetransmitter != NULL, STATUS_NULL_ARG);
+    // freeRtpPacket is idempotent
+    CHK(*ppRetransmitter != NULL, retStatus);
+
+    MEMFREE((*ppRetransmitter)->sequenceNumberList);
+    MEMFREE((*ppRetransmitter)->validIndexList);
+    MEMFREE(*ppRetransmitter);
+CleanUp:
+    CHK_LOG_ERR(retStatus);
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS resendPacketOnNack(PRtcpPacket pRtcpPacket, PKvsPeerConnection pKvsPeerConnection)
+{
+    ENTERS();
+
+    STATUS retStatus = STATUS_SUCCESS;
+    UINT32 senderSsrc = 0, receiverSsrc = 0;
+    UINT32 filledLen = 0, validIndexListLen = 0;
+    PDoubleListNode pCurNode = NULL;
+    PKvsRtpTransceiver pTransceiver, pReceiverTransceiver = NULL, pSenderTranceiver = NULL;
+    UINT64 item;
+    UINT32 index;
+    PRtpPacket pRtpPacket = NULL;
+    PRetransmitter pRetransmitter = NULL;
+
+    CHK(pRtcpPacket != NULL, STATUS_NULL_ARG);
+    CHK_STATUS(rtcpNackListGetSsrcs(pRtcpPacket->payload, pRtcpPacket->payloadLength, &senderSsrc, &receiverSsrc));
+
+    CHK_STATUS(doubleListGetHeadNode(pKvsPeerConnection->pTransceievers, &pCurNode));
+    while(pCurNode != NULL) {
+        CHK_STATUS(doubleListGetNodeData(pCurNode, &item));
+        pTransceiver = (PKvsRtpTransceiver) item;
+
+        if (pTransceiver->jitterBufferSsrc == senderSsrc) {
+            pReceiverTransceiver = pTransceiver;
+        }
+
+        if (pTransceiver->sender.ssrc == receiverSsrc) {
+            pSenderTranceiver = pTransceiver;
+            pRetransmitter = pSenderTranceiver->sender.retransmitter;
+        }
+        pCurNode = pCurNode->pNext;
+    }
+    CHK(pReceiverTransceiver != NULL && pSenderTranceiver != NULL, STATUS_RTCP_INPUT_SSRC_INVALID);
+
+    filledLen = pRetransmitter->seqNumListLen;
+    CHK_STATUS(rtcpNackListGetSeqNums(pRtcpPacket->payload, pRtcpPacket->payloadLength, pRetransmitter->sequenceNumberList, &filledLen));
+    validIndexListLen = pRetransmitter->validIndexListLen;
+    CHK_STATUS(getValidSeqIndexList(pSenderTranceiver->sender.packetBuffer, pRetransmitter->sequenceNumberList,
+            &filledLen, pRetransmitter->validIndexList, &validIndexListLen));
+    for (index = 0; index < validIndexListLen; index++) {
+        retStatus = rollingBufferExtractData(pSenderTranceiver->sender.packetBuffer, pRetransmitter->validIndexList[index], &item);
+        pRtpPacket = (PRtpPacket) item;
+        CHK(retStatus == STATUS_SUCCESS, retStatus);
```

Comment:
It is to safe guard the function as rollingBufferExtractData() has potential to hand over ownership of packet but there is no guarantee all future change on the function would not hand over packet ownership while returning error code, so at least check the result and if it is not null then clean up will free it to avoid memory leak.